// @odoonto7/pacientes-service/src/presentation/http/update-paciente-alergias.http.controller.ts
import {
  Body,
  Controller,
  HttpStatus,
  NotFoundException as NotFoundHttpException,
  Param,
  Patch,
} from '@nestjs/common';
import { routesV1 } from '../../config/app.routes';
import { CommandBus } from '@nestjs/cqrs';
import { match, Result } from 'oxide.ts';
import { ApiOperation, ApiResponse } from '@nestjs/swagger';
import { UpdatePacienteAlergiasCommand } from '../../application/paciente/commands/update-paciente-alergias/update-paciente-alergias.command';
import { UpdatePacienteAlergiasRequestDto } from './commands/update-paciente-alergias.request.dto';
import { AggregateID } from '@odoonto7/shared';
import { NotFoundException } from '@odoonto7/shared';
import { ApiErrorResponse } from '@odoonto7/shared';
import { IdResponse } from '@odoonto7/shared';

@Controller(routesV1.version)
export class UpdatePacienteAlergiasHttpController {
  constructor(private readonly commandBus: CommandBus) {}

  @ApiOperation({ summary: 'Update paciente alergias' })
  @ApiResponse({
    status: HttpStatus.OK,
    type: IdResponse,
  })
  @ApiResponse({
    status: HttpStatus.NOT_FOUND,
    description: NotFoundException.message,
    type: ApiErrorResponse,
  })
  @Patch(`${routesV1.pacientes.root}/:id/alergias`)
  async updateAlergias(
    @Param('id') pacienteId: string,
    @Body() request: UpdatePacienteAlergiasRequestDto,
  ): Promise<IdResponse> {
    const command = new UpdatePacienteAlergiasCommand({
      pacienteId,
      ...request,
    });
    const result: Result<AggregateID, NotFoundException> =
      await this.commandBus.execute(command);

    return match(result, {
      Ok: (id: string) => new IdResponse(id),
      Err: (error: Error) => {
        if (error instanceof NotFoundException)
          throw new NotFoundHttpException(error.message);
        throw error;
      },
    });
  }
}
