// @odoonto7/pacientes-service/src/application/paciente/commands/create-paciente/create-paciente.command.ts
import { Command, CommandProps } from '@odoonto7/shared';
import { Sexo } from '../../../../domain/paciente';

export class CreatePacienteCommand extends Command {
  public readonly nombre: string;
  public readonly apellidos: string;
  public readonly edad: number;
  public readonly sexo: Sexo;
  public readonly telefono?: string;
  public readonly email?: string;
  public readonly alergias?: string;
  public readonly notas?: string;
  public readonly medicacion?: string;
  public readonly patologiasMedicas?: string;
  public readonly embarazada?: boolean;
  public readonly hemorragiasDentales?: boolean;
  public readonly country: string;
  public readonly postalCode: string;
  public readonly street: string;

  constructor(props: CommandProps<CreatePacienteCommand>) {
    super(props);
    this.nombre = props.nombre;
    this.apellidos = props.apellidos;
    this.edad = props.edad;
    this.sexo = props.sexo;
    this.telefono = props.telefono;
    this.email = props.email;
    this.alergias = props.alergias;
    this.notas = props.notas;
    this.medicacion = props.medicacion;
    this.patologiasMedicas = props.patologiasMedicas;
    this.embarazada = props.embarazada;
    this.hemorragiasDentales = props.hemorragiasDentales;
    this.country = props.country;
    this.postalCode = props.postalCode;
    this.street = props.street;
  }
}
