// @odoonto7/pacientes-service/src/modules/paciente.module.ts
import { Module } from '@nestjs/common';
import { CqrsModule } from '@nestjs/cqrs';

// Commands
// import { CreatePacienteCommandHandler } from '../application/paciente/commands/create-paciente/create-paciente.command-handler';
// import { DeletePacienteCommandHandler } from '../application/paciente/commands/delete-paciente/delete-paciente.command-handler';
// import { UpdatePacienteAddressCommandHandler } from '../application/paciente/commands/update-paciente-address/update-paciente-address.command-handler';
// import { UpdatePacienteContactoCommandHandler } from '../application/paciente/commands/update-paciente-contacto/update-paciente-contacto.command-handler';
// import { UpdatePacienteDatosCommandHandler } from '../application/paciente/commands/update-paciente-datos/update-paciente-datos.command-handler';
// import { UpdatePacienteAlergiasCommandHandler } from '../application/paciente/commands/update-paciente-alergias/update-paciente-alergias.command-handler';
// import { UpdatePacienteNotasCommandHandler } from '../application/paciente/commands/update-paciente-notas/update-paciente-notas.command-handler';
// import { UpdatePacienteMedicacionCommandHandler } from '../application/paciente/commands/update-paciente-medicacion/update-paciente-medicacion.command-handler';
// import { UpdatePacientePatologiasCommandHandler } from '../application/paciente/commands/update-paciente-patologias/update-paciente-patologias.command-handler';
// import { UpdatePacienteEmbarazadaCommandHandler } from '../application/paciente/commands/update-paciente-embarazada/update-paciente-embarazada.command-handler';
// import { UpdatePacienteHemorragiasCommandHandler } from '../application/paciente/commands/update-paciente-hemorragias/update-paciente-hemorragias.command-handler';

// Queries
import { FindPacientesQueryHandler } from '../application/paciente/queries/find-pacientes/find-pacientes.query-handler';

// Controllers
import { CreatePacienteHttpController } from '../presentation/http/create-paciente.http.controller';
import { DeletePacienteHttpController } from '../presentation/http/delete-paciente.http-controller';
import { FindPacientesHttpController } from '../presentation/http/find-pacientes.http.controller';
import { UpdatePacienteAddressHttpController } from '../presentation/http/update-paciente-address.http.controller';
import { UpdatePacienteAlergiasHttpController } from '../presentation/http/update-paciente-alergias.http.controller';
import { UpdatePacienteContactoHttpController } from '../presentation/http/update-paciente-contacto.http.controller';
import { UpdatePacienteDatosHttpController } from '../presentation/http/update-paciente-datos.http.controller';
import { UpdatePacienteEmbarazadaHttpController } from '../presentation/http/update-paciente-embarazada.http.controller';
import { UpdatePacienteHemorragiasHttpController } from '../presentation/http/update-paciente-hemorragias.http.controller';
import { UpdatePacienteMedicacionHttpController } from '../presentation/http/update-paciente-medicacion.http.controller';
import { UpdatePacienteNotasHttpController } from '../presentation/http/update-paciente-notas.http.controller';
import { UpdatePacientePatologiasHttpController } from '../presentation/http/update-paciente-patologias.http.controller';

// Infrastructure
import { PacienteRepository } from '../infrastructure/paciente/paciente.repository';
import { PacienteMapper } from '../infrastructure/paciente/paciente.mapper';

// Tokens
export const PACIENTE_REPOSITORY = Symbol('PACIENTE_REPOSITORY');

const commandHandlers = [
  // CreatePacienteCommandHandler,
  // DeletePacienteCommandHandler,
  // UpdatePacienteAddressCommandHandler,
  // UpdatePacienteContactoCommandHandler,
  // UpdatePacienteDatosCommandHandler,
  // UpdatePacienteAlergiasCommandHandler,
  // UpdatePacienteNotasCommandHandler,
  // UpdatePacienteMedicacionCommandHandler,
  // UpdatePacientePatologiasCommandHandler,
  // UpdatePacienteEmbarazadaCommandHandler,
  // UpdatePacienteHemorragiasCommandHandler,
];

const queryHandlers = [FindPacientesQueryHandler];

const controllers = [
  CreatePacienteHttpController,
  DeletePacienteHttpController,
  FindPacientesHttpController,
  UpdatePacienteAddressHttpController,
  UpdatePacienteAlergiasHttpController,
  UpdatePacienteContactoHttpController,
  UpdatePacienteDatosHttpController,
  UpdatePacienteEmbarazadaHttpController,
  UpdatePacienteHemorragiasHttpController,
  UpdatePacienteMedicacionHttpController,
  UpdatePacienteNotasHttpController,
  UpdatePacientePatologiasHttpController,
];

@Module({
  imports: [CqrsModule],
  controllers,
  providers: [
    ...commandHandlers,
    ...queryHandlers,
    PacienteMapper,
    {
      provide: PACIENTE_REPOSITORY,
      useClass: PacienteRepository,
    },
  ],
})
export class PacienteModule {}
