// @odoonto7/pacientes-service/src/presentation/http/queries/find-pacientes.request.dto.ts
import { ApiPropertyOptional } from '@nestjs/swagger';
import { Transform, Type } from 'class-transformer';
import { IsOptional, IsString, IsNumber, Min, Max } from 'class-validator';

export class FindPacientesRequestDto {
  @ApiPropertyOptional({ example: 'Juan', description: 'Nombre del paciente' })
  @IsOptional()
  @IsString()
  readonly nombre?: string;

  @ApiPropertyOptional({ example: 'García', description: 'Apellidos del paciente' })
  @IsOptional()
  @IsString()
  readonly apellidos?: string;

  @ApiPropertyOptional({ example: 'juan@email.com', description: 'Email del paciente' })
  @IsOptional()
  @IsString()
  readonly email?: string;

  @ApiPropertyOptional({ example: 'España', description: 'País de residencia' })
  @IsOptional()
  @IsString()
  readonly country?: string;

  @ApiPropertyOptional({ example: 'Calle Mayor', description: 'Dirección' })
  @IsOptional()
  @IsString()
  readonly street?: string;

  @ApiPropertyOptional({ example: '28001', description: 'Código postal' })
  @IsOptional()
  @IsString()
  readonly postalCode?: string;

  @ApiPropertyOptional({ example: 1, description: 'Página', minimum: 1 })
  @IsOptional()
  @Type(() => Number)
  @IsNumber()
  @Min(1)
  readonly page?: number = 1;

  @ApiPropertyOptional({ example: 20, description: 'Elementos por página', minimum: 1, maximum: 100 })
  @IsOptional()
  @Type(() => Number)
  @IsNumber()
  @Min(1)
  @Max(100)
  readonly limit?: number = 20;
}
